### PR TITLE
Closes #5926: Launch third party app if protocol is not supported by engine.

### LINKED ISSUE #5926: When clicking a link from an Amazon email, Fenix stays blocked on an intent
This is a follow-up to https://github.com/mozilla-mobile/fenix/issues/7331#issuecomment-585745959

Amazon's mails often contain links to the website. When I open this link in Fenix, then the website redirects to an intent. Then the intent doesn't work and Fenix stays blocked here, and as a user I am now blocked.

This works fine in Fennec.

### PR DESCRIPTION
**Pull request by rocketsroger**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -40,6 +40,10 @@ permalink: /changelog/
       * Labels in labeled metrics now have a correct, and slightly stricter, regular expression.
         See [label format](https://mozilla.github.io/glean/user/metrics/index.html#label-format) for more information.
 
+* **feature-app-links**
+  * ⚠️ **This is a breaking change**: `alwaysAllowedSchemes` is removed as a parameter for `AppLinksInterceptor`.
+  * Added `engineSupportedSchemes` as a parameter for `AppLinksInterceptor`.  This allows caller to specifies which protocol is supported by the engine.
+    * Using this information, app links can decide if a protocol should be launched in third party or not regardless of user preference.
```
Comment (Amejia481) [inline docs/changelog.md]: nit:  if a protocol should be launched in **a** third party **app** or not regardless of user preference.

File: docs/changelog.md [hunk 2]
```diff
@@ -40,6 +40,10 @@ permalink: /changelog/
       * Labels in labeled metrics now have a correct, and slightly stricter, regular expression.
         See [label format](https://mozilla.github.io/glean/user/metrics/index.html#label-format) for more information.
 
+* **feature-app-links**
+  * ⚠️ **This is a breaking change**: `alwaysAllowedSchemes` is removed as a parameter for `AppLinksInterceptor`.
+  * Added `engineSupportedSchemes` as a parameter for `AppLinksInterceptor`.  This allows caller to specifies which protocol is supported by the engine.
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply docs/changelog.md]: Agreed.  Thanks for catching this

File: docs/changelog.md (hunk 1 above)
Comment (rocketsroger) [reply docs/changelog.md]: Thanks

File: components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksInterceptorTest.kt [hunk 3]
```diff
@@ -130,26 +128,142 @@ class AppLinksInterceptorTest {
     }
 
     @Test
-    fun `white listed schemes request always intercepted regardless of interceptLinkClicks or launchInApp`() {
+    fun `supported schemes request not launched if launchInApp if false`() {
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksInterceptorTest.kt]: Thanks!

File: components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksInterceptorTest.kt [hunk 4]
```diff
@@ -130,26 +128,142 @@ class AppLinksInterceptorTest {
...
+
+        val supportedUrl = "$supportedScheme://example.com"
+        val supportedRedirect = AppLinkRedirect(Intent.parseUri(supportedUrl, 0), null, null)
+        whenever(mockGetRedirect.invoke(supportedUrl)).thenReturn(supportedRedirect)
+        val response = feature.onLoadRequest(engineSession, supportedUrl, true, false)
+        assertEquals(null, response)
+    }
+
+    @Test
+    fun `supported schemes request not launched if interceptLinkClicks if false`() {
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksInterceptorTest.kt]: Agreed

**Comment by rocketsroger** [comment]
bors r=Amejia481

**Pull request merged**